### ISSUE TITLE
Warn if no explicit memory / storage key word is used

### ISSUE THREAD

**Issue by chriseth** [ethereum/solidity#2525]
This helps #2436.

**Comment by Tectract** [comment]
> Warning: Variable is declared as a storage pointer. Use an explicit \"storage\" keyword to silence this warning.

I have NO IDEA what this is complaining about or how to fix it. The is a BREAKING CHANGE!!! It should not have been accepted without further testing and documentation.

**Comment by axic** [comment]
@Tectract can you please explain the problem? This is a warning and not an error therefore no contract was rendered broken.

Just add the `storage` keyword into the declaration `MyStruct storage mystruct` to silence.

**Comment by Tectract** [comment]
It's a breaking change because contracts that used to compile without error now throw warnings, leading to failed compilation in some systems.

I don't see the point of requiring this keyword.

**Comment by Tectract** [comment]
struct storage notarizedImage {
    string imageURL;
    uint timeStamp;
  }

> ParserError: Expected identifier, got 'Storage'\n struct storage notarizedImage

What? 

Can you please tell me how to fix this contract so it compiles without error?

https://raw.githubusercontent.com/Tectract/ethereum-demo-tools/master/GeektSolidity/contracts/Geekt.sol

**Comment by chriseth** [comment]
Sorry, this is a bug, the compiler also creates warnings for members of structs.

https://github.com/ethereum/solidity/issues/2549

**Comment by Tectract** [comment]
This is a serious bug. I would suggest that this PR be rolled back

I swear to God, it's like you guys are TRYING to make it AS HARD AS POSSIBLE for new-comers to learn solidity. 

You just broke a tutorial that THOUSANDS of people use to learn solidity. 

So NOBODY can tell me how to make this contract compile without warning now?
https://raw.githubusercontent.com/Tectract/ethereum-demo-tools/master/GeektSolidity/contracts/Geekt.sol

That's a serious, serious bug. This contract is being used as a DEMO CONTRACT for a very popular online Solidity tutorial!!! Please revert this change ASAP!

**Comment by Tectract** [comment]
This tool: http://www.enledger.io/EthDeployer/

Can only deploy contracts if there is no warning. You guys just made one of the most popular deployment tools NOT WORK, for most solidity smart contracts.

**Comment by Tectract** [comment]
Ticket opened :
https://github.com/ethereum/solidity/issues/2550

**Comment by axic** [comment]
> It's a breaking change because contracts that used to compile without error now throw warnings, leading to failed compilation in some systems.

Why would compilation fail for a warning? My feeling is that such a "system" should be fixed not to fail.

> I don't see the point of requiring this keyword.

It helps users understand where that member is located (memory vs. storage). In a breaking change it will be mandatory to specify one of the location.

> > ParserError: Expected identifier, got 'Storage'\n struct storage notarizedImage
> What?

It is used at variable declaration and not at struct definition:

```
contract C {
  struct MyStruct {}

  function f() {
    MyStruct storage mystruct = ...;
  }
}
```

> This is a serious bug. I would suggest that this PR be rolled back

The bug @chriseth mentioned is that the warning is issued for struct members.

**Comment by Tectract** [comment]
The problem is that contracts with no error or issues can no longer be compiled without warnings. This is a breaking change in some compilation systems. It is causing me downstream work to get around this BUG.

One major problem is that since solc compilation is no longer included in the node, ALMOST EVERY SINGLE SOLIDITY TUTORIAL was broken. 

I wrote one one how to do it completely client-side, and now my tutorial is also broken. There is no single working Solidity tutorial on the internet right now!!!

**Comment by axic** [comment]
> The problem is that contracts with no error or issues can no longer be compiled without warnings. This is a breaking change in some compilation systems.

Contracts not using structs will not raise any warnings. Once issue #2549 is fixed, even those will work without warnings.

On the other hand can you please explain why is it a breaking change to issue warnings? Warnings are issued for all kinds of things and toolchains should be prepared for this. There is a reason they are called warnings and not errors.

Checked your code at https://github.com/Tectract/EthDeployer/blob/master/src/Deploy.js#L107. You should check if the error list contains a warning. Alternatively using the standard JSON I/O (http://solidity.readthedocs.io/en/develop/using-the-compiler.html#compiler-input-and-output-json-description) via `solc.compileStandardWrapper` (https://github.com/ethereum/solc-js#from-version-0411) should make this extremely easy.

**Comment by Tectract** [comment]
Yes, I need to update EthDeployer, to handle the erroneous case where warnings are thrown in contracts that have no issues. It's downstream work for me and the last, latest Solidity tutorial I wrote (the one that still actually works) will not work for all the readers until you fix the bug upstream or I fix it downstream...

**Comment by axic** [comment]
I beg to differ. EthDeployer doesn't seem to support the case of any warnings, which means EthDeployer contains a bug. Please fix that bug 😉

**Comment by Tectract** [comment]
It specifically shows warnings so you can fix them. Demanding perfection is not a bug. Extra options are always better though. I just have, like, other work today, you know?

**Comment by axic** [comment]
You can still show warnings, yet not fail compilation. Thank you for the bug report, the warning should not be issued on members of structs, that will be fixed!

We're also happy to accept pull requests fixing bugs in Solidity.

**Comment by Tectract** [comment]
> You can still show warnings, yet not fail compilation.

Yes but then deprecated contracts would be deployed, which is not ideal.

Otherwise, I could add another option to deploy "anyways", but then the simplified "one-button-one-click" interface starts to become more complicated...

**Comment by Tectract** [comment]
if(result.errors){

returns true in web3.js even if there are no errors, only warnings?

> You should check if the error list contains a warning.

This is bad. Errors and warnings are different things.

> There is a reason they are called warnings and not errors.

Sigh.

**Comment by chriseth** [comment]
@Tectract can you try compiling directly with solc or solc-js? It is usually way better than web3.js.

**Comment by Tectract** [comment]
I'm not sure I understand the question. I need both web3.js AND solc compiler for the EthDeployer tool, not one or the other.

Since you guys removed solc from the node. I was FORCED to rewrite this tool to use browser-solc. My tutorial is the ONLY working tutorial on the web now, because every single other tutorial used solc from a node connection. Even the ethereum sponsored tutorials and docs have all been broken by that change. However, it's better to use browser-solc, because then you never need anyone to install solc on a server or locally on a client machine. It's actually by far the most elegant solution.

I fixed the issue, although it's still not ideal, now EthDeployer just basically completely ignores warnings, because solc throws erroneous warnings sometimes. It's a degredation in the tool because before you could actually use it to sniff out warnings and it wouldn't deploy if you had any, now it just has to ignore warnings and deprecated contracts could be deployed, possibly. It would take me hours of work to test and update the tool to conditionally allow you to bypass warnings, and it would break the simpified one-button-one-click interface. 

I would implore you guys to no longer accept PR's that can cause erroneous warnings to get returned to the console. That basically breaks the functionality of some tools, such as EthDeployer, because it is supposed to be a simplified debugging platform with only one button.

See: https://github.com/Tectract/EthDeployer/commit/d58af664d79f9459b88669f505f4bd27df33a6a1

and also: https://github.com/ethereum/go-ethereum/issues/3793#issuecomment-314700769

and also: https://blockgeeks.com/guides/how-to-learn-solidity/

**Comment by Tectract** [comment]
Also, in retrospect, it would also be better is if the compiled results object had two sub-objects, results.errors(), and results.warnings(), since, as you mentioned, those really are two different things :)

**Comment by amazingandyyy** [comment]
same issue here here.

**Comment by chriseth** [comment]
@amazingandyyy this should be fixed in 0.4.14 (released some hours ago). Do you still have the problem?
